test_runner/regress/test_replication_start_subxid_overflow.py

```diff
@@ -0,0 +1,217 @@
+def test_replication_start_subxid_overflow3(neon_simple_env: NeonEnv):
+    env = neon_simple_env
+
+    primary = env.endpoints.create_start(branch_name="main", endpoint_id="primary")
+
+    p_con = primary.connect()
+    p_cur = p_con.cursor()
+
+    initdb(p_cur)
+
+    p_cur.execute("show max_connections")
+    max_connections = int(p_cur.fetchall()[0][0])
+    n_connections = max_connections - 3
+    n_subxids = 100
+
+    # Start one top tranaction in primary, with lots of subtransactions. This uses up much the
+    # known-assigned XIDs space in the standby, but doesn't cause it to overflow.
+    large_p_conn = primary.connect()
+    large_p_cur = large_p_conn.cursor()
+    large_p_cur.execute("begin")
+    large_p_cur.execute("select create_subxacts(3000)")
+
+    too_many_xids_error = ".*too many KnownAssignedXids.*"
+    env.pageserver.allowed_errors.extend([too_many_xids_error])
+
+    # Create a replica at this LSN
+    wait_for_last_flush_lsn(env, primary, env.initial_tenant, env.initial_timeline)
+    secondary = env.endpoints.new_replica_start(origin=primary, endpoint_id="secondary")
+    s_con = secondary.connect()
+    s_cur = s_con.cursor()
+
+    # The transaction in primary has not committed yet.
+    wait_replica_caughtup(primary, secondary)
+    s_cur.execute("select count(*) from t")
+    assert s_cur.fetchone() == (0,)
+
+    # Start max number of top transactions in primary, with a lot of
+    # subtransactions each We add the subtransactions to each top
+    # transaction in a round-robin fashion, instead of adding a lot of
+    # subtransactions ot one top transaction at a time. This way, we will
+    # have the max number of subtransactions in the in-memory subxid cache
+    # of each top transaction, until they all overflow.
+    #
+    # Currently, PGPROC_MAX_CACHED_SUBXIDS == 64, so this will overflow the
+    # all the subxid caches after creating 64 subxids in each top
+    # transaction. The point just before the caches have overflowed is the
+    # most interesting point in time, but we'll keep going beyond that, to
+    # ensure that this test is robust even if PGPROC_MAX_CACHED_SUBXIDS
+    # changes.
+    p_cons = []
+    p_curs = []
+    for _ in range(0, n_connections):
+        p_con = primary.connect()
+        p_cur = p_con.cursor()
+        p_cur.execute("begin")
+        p_cons.append(p_con)
+        p_curs.append(p_cur)
+
+    for _subxid in range(0, n_subxids):
+        for i in range(0, n_connections):
+            p_curs[i].execute("select create_subxacts(1)")
+
+    # Commit all the transactions in the primary
+    large_p_cur.execute("commit")
+    for i in range(0, n_connections):
+        p_curs[i].execute("commit")
+
+    time.sleep(5)  # wait for replica to catch-up
+
+    log_path = secondary.endpoint_path() / "compute.log"
+    with open(log_path, "r") as log_file:
+        replica_log = log_file.read()
+        assert "too many KnownAssignedXids" in replica_log
```

Comment:
You could use the `assert_log_contains` function for this

Reply:
`assert_log_contains` is available only for page server and broker

Reply:
Hmm, it would be straightforward to make it work for Endpoints too, I think

Reply:
+1
Just didn't wan to add more changes to this PR.
And actually this is not the first place where compute log is parsed in this way...
So in case of such refactoring of fixtures classes we also need  to update some other tests.
And this is why it is reasonable to perform it as separate PR.

Reply:
Opened https://github.com/neondatabase/neon/pull/7997 for this